coderedcms/wagtail_flexible_forms/blocks.py

```diff
@@ -0,0 +1,269 @@
+from django import forms
+from django.db.models import BLANK_CHOICE_DASH
+from django.utils.dateparse import parse_datetime
+from django.utils.encoding import force_text
+from django.utils.text import slugify
+from django.utils.translation import ugettext_lazy as _
+from unidecode import unidecode
+from wagtail.core.blocks import (
+    StructBlock, TextBlock, CharBlock, BooleanBlock, ListBlock, StreamBlock,
+    DateBlock, TimeBlock, DateTimeBlock, ChoiceBlock, RichTextBlock,
+)
+
+
+class FormFieldBlock(StructBlock):
+    field_label = CharBlock(label=_('Label'))
+    help_text = TextBlock(required=False, label=_('Help text'))
+
+    field_class = forms.CharField
+    widget = None
+
+    def get_slug(self, struct_value):
+        return force_text(slugify(unidecode(struct_value['field_label'])))
+
+    def get_field_class(self, struct_value):
+        return self.field_class
+
+    def get_widget(self, struct_value):
+        return self.widget
+
+    def get_field_kwargs(self, struct_value):
+        kwargs = {'label': struct_value['field_label'],
+                  'help_text': struct_value['help_text'],
+                  'required': struct_value.get('required', False)}
+        if 'default_value' in struct_value:
+            kwargs['initial'] = struct_value['default_value']
+        form_widget = self.get_widget(struct_value)
+        if form_widget is not None:
+            kwargs['widget'] = form_widget
+        return kwargs
+
+    def get_field(self, struct_value):
+        return self.get_field_class(struct_value)(
+            **self.get_field_kwargs(struct_value))
+
+
+class OptionalFormFieldBlock(FormFieldBlock):
+    required = BooleanBlock(label=_('Required'), required=False)
+
+
+CHARFIELD_FORMATS = [
+    ('email', _('Email')),
+    ('url', _('URL')),
+]
+try:
+    from phonenumber_field.formfields import PhoneNumberField
+except ImportError:
+    pass
+else:
+    CHARFIELD_FORMATS.append(('phone', _('Phone')))
+
+
+class CharFieldBlock(OptionalFormFieldBlock):
+    format = ChoiceBlock(choices=CHARFIELD_FORMATS, required=False, label=_('Format'))
+    default_value = CharBlock(required=False, label=_('Default value'))
+
+    class Meta:
+        label = _('Text field (single line)')
+
+    def get_field_class(self, struct_value):
+        text_format = struct_value['format']
+        if text_format == 'url':
+            return forms.URLField
+        if text_format == 'email':
+            return forms.EmailField
+        if text_format == 'phone':
+            return PhoneNumberField
+        return super().get_field_class(struct_value)
```

Comment:
Interestingly enough, these fields are not rendering with the correct HTML5 input type (email, tel). They are all rendering as `type="text"`. Not sure if that's a problem here, in the bootstrap4 package, or elsewhere.